test_runner/regress/test_layer_writers_fail.py

```diff
@@ -0,0 +1,100 @@
+import pytest
+from fixtures.neon_fixtures import NeonEnv
+
+
+def test_image_layer_writer_fail_before_finish(neon_simple_env: NeonEnv):
+    env = neon_simple_env
+    pageserver_http = env.pageserver.http_client()
+
+    tenant_id, _ = env.neon_cli.create_tenant(
+        conf={
+            # small checkpoint distance to create more delta layer files
+            "checkpoint_distance": f"{1024 ** 2}",
+            # set the target size to be large to allow the image layer to cover the whole key space
+            "compaction_target_size": f"{1024 ** 3}",
+            # tweak the default settings to allow quickly create image layers and L1 layers
+            "compaction_period": "1 s",
+            "compaction_threshold": "2",
+            "image_creation_threshold": "1",
+        }
+    )
+    timeline_id = env.neon_cli.create_timeline(
+        "test_image_layer_writer_fail_before_finish", tenant_id=tenant_id
+    )
+    pg = env.postgres.create_start(
+        "test_image_layer_writer_fail_before_finish", tenant_id=tenant_id
+    )
+
+    pg.safe_psql_many(
+        [
+            "CREATE TABLE foo (t text) WITH (autovacuum_enabled = off)",
+            """INSERT INTO foo
+        SELECT 'long string to consume some space' || g
+        FROM generate_series(1, 100000) g""",
+        ]
+    )
+
+    pageserver_http.configure_failpoints(("image-layer-writer-fail-before-finish", "return"))
+    with pytest.raises(Exception, match="image-layer-writer-fail-before-finish"):
+        pageserver_http.timeline_checkpoint(tenant_id, timeline_id)
+
+    new_temp_layer_files = list(
+        filter(
+            lambda file: str(file)[-7:] == "___temp",
+            [path for path in env.timeline_dir(tenant_id, timeline_id).iterdir()],
+        )
+    )
+
+    assert (
+        len(new_temp_layer_files) == 0
+    ), "pageserver should clean its temporary new image layer files on failure"
+
+
+def test_delta_layer_writer_fail_before_finish(neon_simple_env: NeonEnv):
+    env = neon_simple_env
+    pageserver_http = env.pageserver.http_client()
+
+    tenant_id, _ = env.neon_cli.create_tenant(
```

Comment:
Both tests could run on `main` branch  (it has id `_` here) with the same effect, but less code.
Your test case it not really related to branching, so might be simpler to remove it.